modules/templates-writing-parameters.adoc

```diff
@@ -63,13 +63,14 @@ characters long consisting of all upper and lowercase alphabet letters
 and numbers.
 
 The syntax available is not a full regular expression syntax. However, you can
-use `\w`, `\d`, and `\a` modifiers:
+use `\w`, `\d`, `\a`, and `\A` modifiers:
 
 - `[\w]{10}` produces 10 alphabet characters, numbers, and underscores. This
 follows the PCRE standard and is equal to `[a-zA-Z0-9_]{10}`.
 - `[\d]{10}` produces 10 numbers. This is equal to `[0-9]{10}`.
 - `[\a]{10}` produces 10 alphabetical characters. This is equal to
 `[a-zA-Z]{10}`.
+- `[\A]{10}` produces 10 non-alphabetical characters.
```

Comment:
Same as above `[\\w]{10}`, `[\\d]{10}`, `[\\a]{10}`, `[\\A]{10}`

Reply:
`[\A]` will expand to any of:

    ~!@#$%^&*()-_+={}[]\|<,>.?/"';:`

I think the PCRE equivalent is:

    [~!@#$%\^&*()\-_+={}\[\]\\|<,>.?/"';:`]

... but in order to interpret that the reader has to have an understanding of the special escaping rules within bracketed character classes. For simplicity I chose to prefix `\` `^` `-` `[` and `]` with a backslash, even though some of them don't strictly require it, and others don't require it if the order of the characters within the character class are shuffled to make them not trigger their special meanings.

Maybe you could get away with saying it's equivalent to `[[:punct:]]`, but I suppose what matches that depends on locale etc...

Reply:
Thanks for the guidance @xiuwang and Sam. I've updated the documentation and added a note to clarify the `[\\A]` example. Please let me know if it's clear.

https://deploy-preview-30790--osdocs.netlify.app/openshift-enterprise/latest/openshift_images/using-templates.html#templates-writing-parameters_using-templates

Reply:
The markdown rendering looks a bit weird:

![image](https://user-images.githubusercontent.com/309468/112334876-ae863e00-8cb3-11eb-80e7-ef7f50156de9.png)

And the PCRE-format example should have `[10]` on the end of it.

You could say, "`[\A]{10}` produces 10 punctuation characters" rather than the odd sounding "non-alphabetical".

To be honest I'm not sure why the double-backslash is requested by @xiuwang - the extra backslash is making it through to the rendered documentation, this looks wrong to me?

Reply:
[\A]{10} does't work when instantiating the template file
` grep "A]" cakephp-mysql-persistent.json  -rn --color`
`562:			"from": "[\A]{10}"`

`oc new-app cakephp-mysql-persistent.json `
`W0401 10:47:04.436294   16528 dockerimagelookup.go:237] container image remote registry lookup failed: unable to parse docker reference : invalid reference format`
`error: unable to load template file "cakephp-mysql-persistent.json": error parsing cakephp-mysql-persistent.json: json: line 561: invalid character 'A' in string escape code`

In https://github.com/openshift/cluster-samples-operator/blob/master/assets/operator/ocp-x86_64/cakephp/templates/cakephp-mysql-persistent.json#L569 , it used double-backslash to generate values.

Reply:
Ah, you're talking about escaping the backslash character within a JSON string. But the template language itself interprets `\A`. I've written templates mostly using YAML with unquoted strings, where backslashes don't need to be escaped, hence my confusion. :)

I think it would be better to document these patterns with a single backslash - as that's what the template processor actually expects. But if you want to make it clear, add a note saying that the backslash may need to be escaped depending on whether the template is being written in JSON or YAML, and the type of string that it's embedded within. For example these are all equivalent:

```
[... YAML boilerplate...]
parameters:
- name: unquoted_example
  generate: expression
  from: [\A]{10}
- name: singlequoted_example
  generate: expression
  from: '[\A]{10}'
- name: doublequoted_example
  generate: expression
  from: "[\\A]{10}"
```

and

```
{
[...  JSON boilerplate ...]
{"parameters": [
  {"name": "json_example",
   "generate": "expression",
   "from": "[\\A]{10}"
  }
}
```

Reply:
Sam, Thanks, you are correct. I only force on the json format.
@EricPonvelle Agreed with Sam, we should use a single backslash to document.